[DCJ-511] azureTableThreadpool blocks when saturated

okotsopoulos opened the pull request:
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DCJ-511

## Addresses

TDR and other services use [ThreadPoolTaskExecutor](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/scheduling/concurrent/ThreadPoolTaskExecutor.html)s to parallelize tasks, which are comprised of a core pool of threads for processing and a queue as a fallback for when those threads are active. If the queue fills up, tasks submitted to the executor will be rejected with a [RejectedExecutionException](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/RejectedExecutionException.html).

Making the queue bigger or even unbounded runs the risk of memory issues if requests come faster than they can be processed. In some places, TDR tries to batch up its submissions to an executor to avoid overloading it, only proceeding to the next batch when the previous has succeeded. But this does not make the most efficient use of the executor's capacity, nor does it guard against the scenario where many separate requests are submitting tasks the same executor.

Our driver problem here is that Nate's Azure ingests with 10s of thousands of file references are still encountering intermittent failures due to threadpool saturation, despite increasing the queue size to 10,000.

## Summary of changes

- Introduce new `BlockingRejectedExecutionHandler` to override executor behavior.  If an executor cannot accept a task due to saturation of its threadpool and queue, this handler blocks when waiting for a spot to become available in the queue rather than rejecting the task.
- Set `azureTableThreadpool`'s rejected execution handler to a new instance of `BlockingRejectedExecutionHandler` if the threadpool is configured to block when saturated.
- Set application property `azure.threading.blockWhenSaturated=true` -- can be overridden at the environment level by setting environment variable `AZURE_THREADING_BLOCKWHENSATURATED`.

## Testing Strategy

- Expanded unit testing.

## Future Opportunities

- Evaluate TDR's other existing `ThreadPoolTaskExecutor`s to see if blocking handling is a good fit for them, and wire them in if so.
- Convert TDR's older `ThreadPoolExecutor`s to `ThreadPoolTaskExecutor`s for automatic threadpool and queue instrumentation via Micrometer / Actuator, then evaluate for blocking handling.
- In places where TDR presently batches submissions to a threadpool to avoid saturation, refactor out in favor of blocking handling.

<!-- Reminder -->
<!-- Two CODEOWNERS will be automatically assigned to review this pull request. If you otherwise have two reviewers, you do not need to wait for their review. -->

okotsopoulos committed:
[DCJ-511] Introduce BlockingRejectedExecutionHandler

This can be supplied to ThreadPoolExecutors to define behavior when a task can't be accepted.

If the denial is due to saturation of the threadpool and queue, this handler blocks when waiting for a spot in the queue.

The present default behavior is that the task is rejected via a thrown RejectedExecutionException.

okotsopoulos committed:
azureTableThreadpool blocks when saturated

This is dictated by property azure.threading.blockWhenSaturated.

okotsopoulos committed:
Add unit tests for BlockingRejectedExecutionHandler

okotsopoulos commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceConfiguration.java:
```diff
@@ -173,6 +174,9 @@ public AsyncTaskExecutor azureTableThreadpool() {
     executor.setKeepAliveSeconds(0);
     executor.setQueueCapacity(threading().maxQueueSize());
     executor.setThreadNamePrefix("az-table-thread-");
+    if (threading().blockWhenSaturated()) {
```
We may ultimately wish to remove this config switch, but at least when rolling out this new behavior I wanted to retain the ability to revert to the default rejection behavior.

okotsopoulos commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceConfiguration.java:
```diff
@@ -163,7 +164,7 @@ public record Synapse(
-  public record Threading(int numTableThreads, int maxQueueSize) {}
+  public record Threading(int numTableThreads, int maxQueueSize, boolean blockWhenSaturated) {}
 
   @Bean("azureTableThreadpool")
```
All of TDR's usages of this threadpool are within Stairway flight execution, so we favor blocking on saturation rather than failing fast.  If we were using a threadpool as part of servicing synchronous API responses, then we may favor the default rejection strategy.  I'll keep this in mind as I evaluate other existing threadpools.

okotsopoulos commented on src/main/java/bio/terra/common/BlockingRejectedExecutionHandler.java:
```diff
@@ -0,0 +1,55 @@
+    if (!executor.isShutdown()) {
+      try {
+        // Block until space becomes available in the queue
+        executor.getQueue().put(r);
```
Note that `ThreadPoolExecutor`s use a `BlockingQueue`:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html#getQueue--

Calling `put` on a `BlockingQueue` blocks until space is available:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html#put-E-

okotsopoulos commented on src/main/java/bio/terra/common/BlockingRejectedExecutionHandler.java:
Thanks @pshapiro4broad for the inspiration for this approach :) I was previously pursuing a Semaphore-based wrapper to our executors.

okotsopoulos committed:
Driveby - store azureTableThreadpool name in static variable

This way, we can easily see in our IDE all the places where this Spring Bean is being pulled into another component.

We no longer have to redefine the same string throughout code and tests.

pshapiro4broad commented on src/main/java/bio/terra/common/BlockingRejectedExecutionHandler.java:
```diff
@@ -0,0 +1,55 @@
+public class BlockingRejectedExecutionHandler implements RejectedExecutionHandler {
+  @VisibleForTesting
+  static String interruptedExceptionMessage(Runnable r, ThreadPoolExecutor executor) {
+    return "Task %s interrupted while waiting to be added to queue: %s".formatted(r, executor);
```
Does it make sense to use `Runnable.toString()` here? By default this is the same as `Object.toString()` (which would print the object's type and memory address), but maybe the `Runnable`s used in TDR will print something more useful.

okotsopoulos replied:
Good shout -- I don't think this is useful, so I'll remove it.  Conversely, the `ThreadPoolExecutor.toString()` method is very useful so I'll leave that in:

https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/concurrent/ThreadPoolExecutor.java#L1856-L1891

pshapiro4broad commented on src/main/java/bio/terra/service/resourcemanagement/azure/AzureResourceConfiguration.java:
```diff
@@ -163,16 +164,25 @@ public record Synapse(
+   * The name of a {@link AsyncTaskExecutor} Spring Bean which executes tasks related to Azure
+   * Storage Tables.
+   */
+  public static final String TABLE_THREADPOOL_NAME = "azureTableThreadpool";
```
This is fine, but I wonder if having a subclass (or wrapper class) would be easier to use than a qualifier. For example, a subclass could be written as:
```
@Component
public class AzureTableThreadPool extends ThreadPoolTaskExecutor {

  public AzureTableThreadPool(AzureResourceConfiguration azureResourceConfiguration) {
    var threading = azureResourceConfiguration.threading();

    setCorePoolSize(threading.numTableThreads());
    setMaxPoolSize(threading.numTableThreads());
    setKeepAliveSeconds(0);
    setQueueCapacity(threading.maxQueueSize());
    setThreadNamePrefix("az-table-thread-");
    if (threading.blockWhenSaturated()) {
      executor.setRejectedExecutionHandler(new BlockingRejectedExecutionHandler());
    }
    initialize();
  }
}
```

Then you could inject using `AzureTableThreadPool` and avoid the need for `@Qualifier`

okotsopoulos replied:
I liked this suggestion and started to pursue it, but found that changes to tests it required make it more than I want to take on within this PR.

pshapiro4broad reviewed (approved):
This looks OK to me, it'll be interesting to see how it works in practice.

After reading the summary, I did wonder about using `RetryRuleExponentialBackoff` instead, if we're trying to fix an issue inside a stairway flight execution. My thinking is that stairway itself already has a queuing mechanism, and using it to queue would avoid TDR running out of memory as more and more threads are blocked waiting on the `put()` operation. Although maybe stairway itself doesn't support a retry at this stage of operation?

rushtong reviewed (approved):
This looks reasonable to me, thank you!

fboulnois reviewed (approved):
looks good 👍

okotsopoulos commented:
> After reading the summary, I did wonder about using RetryRuleExponentialBackoff instead, if we're trying to fix an issue inside a stairway flight execution. My thinking is that stairway itself already has a queuing mechanism, and using it to queue would avoid TDR running out of memory as more and more threads are blocked waiting on the put() operation. Although maybe stairway itself doesn't support a retry at this stage of operation?

Hi @pshapiro4broad , Stairway's retry mechanism wouldn't help us recover here because it's equipped to retry a failed step, and the instances of threadpool saturation that have been manifesting themselves recently have been full saturation within the execution of a single step.

Stairway's queuing mechanism exists at the flight level rather than the step level.  When a flight is consumed from the queue and executed, it does so on the same thread for the duration of its execution attempt.  The queue does not come into play when walking through its steps.

pshapiro4broad commented:
> Hi @pshapiro4broad , Stairway's retry mechanism wouldn't help us recover here because it's equipped to retry a failed step, and the instances of threadpool saturation that have been manifesting themselves recently have been full saturation within the execution of a single step.
>
> Stairway's queuing mechanism exists at the flight level rather than the step level. When a flight is consumed from the queue and executed, it does so on the same thread for the duration of its execution attempt. The queue does not come into play when walking through its steps.

Oh, I wasn't aware that stairway's step retry mechanism didn't cause the current flight to be queued. I had assumed that it used its queuing mechanism to handle retry backoffs and not a sleep() directly on the flight's execution thread.

But even if this did exist, it would probably have other downsides since the step would fail in the middle and then restart from the beginning of the step.

okotsopoulos committed:
Remove Stringified Runnable from error messages

They are unlikely to provide useful information to developers or users.

From @pshapiro4broad PR feedback --

>Does it make sense to use Runnable.toString() here? By default this is the same as Object.toString() (which would print the object's type and memory address), but maybe the Runnables used in TDR will print something more useful.

okotsopoulos committed:
Merge branch 'develop' into okotsopo-DCJ-511-blocking-rejectedexecutionhandler
